src/main/java/ch/njol/skript/expressions/ExprSignText.java

```diff
@@ -32,145 +26,203 @@
+				return new String[0];
+			int index = lineNumber.intValue() - 1;
+			if (event instanceof SignChangeEvent && getTime() >= 0 && block == ((SignChangeEvent) event).getBlock() && !delayed) {
+				return CollectionUtils.array(((SignChangeEvent) event).getLine(index));
+			}
+			if (block == null || !sign.isOfType(block)) {
+				return new String[0];
+			}
+			return CollectionUtils.array(
+				((Sign) (block.getState())).getLine(index)
+			);
 		}
-		final Block b = block.getSingle(e);
-		if (b == null)
-			return new String[0];
-		if (!sign.isOfType(b))
-			return new String[0];
-		return new String[] {((Sign) b.getState()).getLine(line)};
 	}
-	
-	@Override
-	public String toString(final @Nullable Event e, final boolean debug) {
-		return "line " + line.toString(e, debug) + " of " + block.toString(e, debug);
-	}
-	
-	// TODO allow add, remove, and remove all (see ExprLore)
+
 	@Override
 	@Nullable
 	public Class<?>[] acceptChange(final ChangeMode mode) {
-		if (mode == ChangeMode.DELETE || mode == ChangeMode.SET)
-			return new Class[] {String.class};
+		if (isLines) {
+			if (mode == ChangeMode.RESET) {
+				return null;
+			}
+			return CollectionUtils.array(String[].class);
+		} else if (mode == ChangeMode.SET || mode == ChangeMode.DELETE) {
+			return CollectionUtils.array(String.class);
+		}
 		return null;
 	}
-	
-	static boolean hasUpdateBooleanBoolean = true;
-	
-	@SuppressWarnings("incomplete-switch")
+
 	@Override
-	public void change(final Event e, final @Nullable Object[] delta, final ChangeMode mode) throws UnsupportedOperationException {
-		final Number l = line.getSingle(e);
-		if (l == null)
-			return;
-		final int line = l.intValue() - 1;
-		if (line < 0 || line > 3)
-			return;
-		final Block b = block.getSingle(e);
-		if (b == null)
+	public void change(Event event, Object @Nullable [] delta, ChangeMode mode) {
+		if (delta == null) {
+			if (mode != ChangeMode.DELETE && mode != ChangeMode.RESET)
+				return;
+		}
+		Block block = exprBlock.getSingle(event);
+		if (block == null)
 			return;
-		if (getTime() >= 0 && e instanceof SignChangeEvent && b.equals(((SignChangeEvent) e).getBlock()) && !Delay.isDelayed(e)) {
+		if (isLines) {
+			if (!sign.isOfType(block))
+				return;
+			Sign state = (Sign) block.getState();
 			switch (mode) {
+				case ADD:
+					for (String text : (String[]) delta) {
+						int index = 0;
+						while (!state.getLine(index).equals("")) {
+							index++;
+							if (index > 3)
+								break;
+						}
+						if (index <= 3)
+							state.setLine(index, Utils.replaceChatStyles(text));
+					}
+					break;
 				case DELETE:
-					((SignChangeEvent) e).setLine(line, "");
+					for (int index = 0; index < 4; index++) {
+						state.setLine(index, "");
+					}
 					break;
-				case SET:
-					assert delta != null;
-					((SignChangeEvent) e).setLine(line, (String) delta[0]);
+				case REMOVE:
+					for (String text : (String[]) delta) {
+						for (int index = 0; index <= 3; index++) {
```

Comment:
Keep inline with your delete changemode switch option
```suggestion
						for (int index = 0; index < 4; index++) {
```